libqtile/widget/bluetooth.py

```diff
@@ -46,70 +52,50 @@ def __init__(self, **config):
-                                                signature='ss',
-                                                body=['org.bluez.Adapter1', 'Powered'])
-        if power_msg.message_type == MessageType.METHOD_RETURN:
-            self.powered = power_msg.body[0].value if power_msg.body else False
-        else:
-            logger.warning('Failed to get bluetooth power status.')
-            self.powered = False
+        # set initial values
+        self.powered = await self._init_adapter()
+        self.connected, self.device = await self._init_device()
 
         self.update_text()
 
-        # add callbacks for adapter and device
+        # add receiver routines for adapter and device
         subscribed_adapter = await add_signal_receiver(self._signal_received_adapter,
                                                        session_bus=False,
                                                        signal_name='PropertiesChanged',
-                                                       path='/org/bluez/hci0',
-                                                       dbus_interface='org.freedesktop.DBus.Properties')
+                                                       path=BLUEZ_PATH,
+                                                       dbus_interface=BLUEZ_PROPERTIES)
         if not subscribed_adapter:
             logger.warning('Could not subscribe to bluez adapter.')
 
         subscribed_device = await add_signal_receiver(self._signal_received_device,
                                                       session_bus=False,
                                                       signal_name='PropertiesChanged',
-                                                      path='/org/bluez/hci0/' + self.hci,
-                                                      dbus_interface='org.freedesktop.DBus.Properties')
+                                                      path=BLUEZ_PATH + self.hci,
+                                                      dbus_interface=BLUEZ_PROPERTIES)
         if not subscribed_device:
             logger.warning('Could not subscribe to bluez device.')
 
+    async def _init_adapter(self):
+        # set up interface to adapter properties using high-level api
+        bus = await MessageBus(bus_type=BusType.SYSTEM).connect()
+        introspect = await bus.introspect(BLUEZ, BLUEZ_PATH)
+        obj = bus.get_proxy_object(BLUEZ, BLUEZ_PATH, introspect)
+        iface = obj.get_interface(BLUEZ_PROPERTIES)
+
+        powered = await iface.call_get(BLUEZ_ADAPTER, 'Powered')
+        return powered.value
+
+    async def _init_device(self):
+        # set up interface to device properties using high-level api
+        bus = await MessageBus(bus_type=BusType.SYSTEM).connect()
+        introspect = await bus.introspect(BLUEZ, BLUEZ_PATH + self.hci)
+        obj = bus.get_proxy_object(BLUEZ, BLUEZ_PATH + self.hci, introspect)
+        iface = obj.get_interface(BLUEZ_PROPERTIES)
```

Comment:
You don't actually need to use the properties interface as dbus-next exposes the properties in the object. So you can do:
```
iface = obj.get_interface(BLUZ_DEVICE)
connected = await iface.get_connected()
name = await iface.get_name()
```
The syntax for retrieving properties is `get_` + property name in snake case. Properties can be set with `set_`.

Reply:
That's definitely easier then, will add that shortly

Reply:
Look good?